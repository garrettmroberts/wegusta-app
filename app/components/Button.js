import React from 'react';
import { View, Text, Pressable, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

import styleConstants from '../config/index';

const Button = props => {
  return (
    <Pressable { ...props }
      style={ ({pressed}) => [
        {
          opacity: pressed ? 0.8 : 1
        },
        styles.buttonStyle
      ]}
    >
      <View style={ styles.buttonStyle }>
        <Text style={ styleConstants.fonts.button }>{ props.text }</Text>
      </View>
    </Pressable>
  );
};

Button.propTypes = {
  text: PropTypes.string.isRequired,
  handlePressIn: PropTypes.func,
  handlePressOut: PropTypes.func
};

const styles = StyleSheet.create({
  buttonStyle: {
    backgroundColor: styleConstants.colors.primary,
    height: 48,
    width: styleConstants.constants.screenWidth - 48,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 12
  }
});

export default Button;
