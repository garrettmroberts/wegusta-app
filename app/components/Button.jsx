import React from 'react';
import { View, Text, Pressable, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import { Feather } from '@expo/vector-icons';

import { colors, fonts, constants } from '../config';

const Button = props => {
  let content; 
  if (props.type === 'card') {
    content = <Feather name="feather" size={40} color={colors.white} />;
  } else if (props.type === 'icon') {
    content = (
      <View>
        <Feather name="plus" size={20} color={colors.white} />
        <Text style={fonts.button}>{props.text || '' }</Text>
      </View>
    );
  } else {
    content = <Text style={ fonts.button }>{ props.text || '' }</Text>;
  }

  return (
    <Pressable { ...props }
      style={ ({pressed}) => [
        {
          opacity: pressed ? 0.8 : 1
        }
      ]}
    >
      <View style={[
        props.type === 'default' || !props.type ? styles.buttonDefault : '',
        props.type === 'card' ? styles.buttonCard : '',
        props.type === 'icon' ? styles.buttonIcon : ''
      ]}>
        { content }
      </View>
    </Pressable>
  );
};

Button.propTypes = {
  text: PropTypes.string, // Required when type='default' or type='icon'
  type: PropTypes.string,
  handlePress: PropTypes.func
};

const styles = StyleSheet.create({
  buttonDefault: {
    backgroundColor: colors.primary,
    height: 48,
    width: constants.screenWidth - 48,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 12
  },
  buttonCard: {
    height: 80,
    width: 80,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 40,
    margin: 16,
    shadowColor: colors.black,
    shadowOffset: {
      width: 0,
      height: 2
    },
    shadowOpacity: 0.23,
    shadowRadius: 2.62,
    elevation: 4
  },
  buttonIcon: {
    backgroundColor: colors.black,
    height: 32,
    width: 96,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    borderRadius: 12,
    marginRight: 20
  }
});

export default Button;
