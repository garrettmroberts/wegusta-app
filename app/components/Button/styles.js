import { StyleSheet } from 'react-native';
import { colors, fonts } from '../../config';

const styles = StyleSheet.create({
  centered: {
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    alignItems: 'stretch'
  },
  stretch: {
    flex: 1
  },
  contentWrapper: {
    justifyContent: 'center',
    alignItems: 'center',
    flexWrap: 'nowrap',
    flexDirection: 'row'
  },
  buttonSmall: {
    borderRadius: 12
  },
  buttonReg: {
    borderRadius: 8
  },
  primary: {
    backgroundColor: colors.buttonPrimary
  },
  secondary: {
    backgroundColor: colors.buttonSecondary
  },
  tertiary: {
    backgroundColor: colors.white
  },
  disabled: {
    backgroundColor: colors.buttonSecondary
  },
  destructive: {
    backgroundColor: colors.buttonDestructive
  },
  fullWidth: {
    height: 48,
    width: '100%',
    paddingHorizontal: 24
  },
  large: {
    height: 48,
    paddingHorizontal: 32
  },
  medium: {
    height: 40,
    paddingHorizontal: 32
  },
  small: {
    height: 32,
    paddingHorizontal: 32
  },
  text: {
    marginHorizontal: 10,
    fontWeight: fonts.button.fontWeight,
    fontSize: fonts.button.fontSize,
    lineHeight: fonts.button.lineHeight,
    alignSelf: 'center'
  },
  darkText: {
    color: colors.black
  },
  grayText: {
    color: colors.disabledGray
  }
});

export default styles;
