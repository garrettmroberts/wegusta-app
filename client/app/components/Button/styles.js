import { StyleSheet } from 'react-native'
import { colors, constants } from '../../config'

const styles = StyleSheet.create({
    centered: {
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'row',
        alignItems: 'stretch',
    },
    contentWrapper: {
        justifyContent: 'center',
        alignItems: 'center',
        flexWrap: 'nowrap',
        flexDirection: 'row',
    },
    buttonSmall: {
        borderRadius: 8,
    },
    buttonReg: {
        borderRadius: 12,
    },
    primary: {
        backgroundColor: colors.primary,
    },
    secondary: {
        backgroundColor: colors.greyLight,
    },
    tertiary: {
        backgroundColor: colors.white,
    },
    disabled: {
        backgroundColor: colors.greyLight,
    },
    destructive: {
        backgroundColor: colors.error,
    },
    fullWidth: {
        height: 56,
        width: constants.screenWidth - 32,
        paddingHorizontal: 24,
    },
    large: {
        height: 48,
        paddingHorizontal: 32,
    },
    medium: {
        height: 40,
        paddingHorizontal: 32,
    },
    small: {
        height: 32,
        paddingHorizontal: 32,
    },
    text: {
        marginHorizontal: 10,
        alignSelf: 'center',
    },
    darkText: {
        color: colors.primary,
    },
    grayText: {
        color: colors.grey,
    },
    lightText: {
        color: colors.white,
    },
})

export default styles
