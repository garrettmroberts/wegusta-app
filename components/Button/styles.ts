import { StyleSheet } from 'react-native'
import Colors from '../../constants/Colors'
import Sizes from '../../constants/Sizes'

const styles = StyleSheet.create({
  centered: {
    alignItems: 'stretch',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  contentWrapper: {
    justifyContent: 'center',
    alignItems: 'center',
    flexWrap: 'nowrap',
    flexDirection: 'row',
  },
  buttonSmall: {
    borderRadius: 8,
  },
  buttonReg: {
    borderRadius: 12,
  },
  primary: {
    backgroundColor: Colors.primary,
  },
  secondary: {
    backgroundColor: Colors.greyLight,
  },
  tertiary: {
    backgroundColor: Colors.white,
  },
  disabled: {
    backgroundColor: Colors.greyLight,
  },
  destructive: {
    backgroundColor: Colors.error,
  },
  fullWidth: {
    height: 56,
    width: Sizes.screenWidth - 32,
    paddingHorizontal: 24,
  },
  large: {
    height: 48,
    paddingHorizontal: 32,
  },
  medium: {
    height: 40,
    paddingHorizontal: 32,
  },
  small: {
    height: 32,
    paddingHorizontal: 32,
  },
  text: {
    marginHorizontal: 10,
    alignSelf: 'center',
  },
  darkText: {
    color: Colors.primary,
  },
  grayText: {
    color: Colors.grey,
  },
  lightText: {
    color: Colors.white,
  },
})

export default styles