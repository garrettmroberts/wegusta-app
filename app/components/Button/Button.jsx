import React from 'react';
import { View, Text, Pressable, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

import styles from './styles';
import { fonts } from '../../config';

const getTextStyle = (type) => {
  if (type === 'primary' || type === 'destructive') {
    return [fonts.button, styles.text];
  } else if (type === 'secondary' || type === 'tertiary') {
    return [fonts.button, styles.darkText, styles.text];
  } else {
    return [fonts.button, styles.grayText, styles.text];
  }
};

const Button = ({ type, size, text, iconBefore, iconAfter, handlePress }) => {
  return (
    <Pressable style={({ pressed }) => [
      styles.button, 
      styles[type], 
      styles[size], 
      { opacity: pressed ? 0.8 : 1} ]} 
    onPress={ handlePress } 
    >
      { iconBefore }
      <Text style={ getTextStyle(type) }>{ text }</Text>
      { iconAfter }
    </Pressable>
  );
};

Button.propTypes = {
  type: PropTypes.string.isRequired,
  size: PropTypes.string.isRequired,
  text: PropTypes.string,
  iconBefore: PropTypes.element,
  iconAfter: PropTypes.element,
  handlePress: PropTypes.func

};

export { Button };
