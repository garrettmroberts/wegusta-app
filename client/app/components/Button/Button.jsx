import React from 'react'
import { Text, View, Pressable } from 'react-native'

import PropTypes from 'prop-types'

import styles from './styles'
import { fonts, colors } from '../../config'
import buildIcon from '../../utils/buildIcon'

const Button = ({
  type,
  size,
  icon,
  iconPlacement,
  text,
  style,
  handlePress,
}) => {
  const getTextStyle = (type, size) => {
    let classlist = [styles.text]

    classlist.push(size === 'small' ? fonts.buttonSmall : fonts.buttonReg)
    if (type === 'primary' || type === 'destructive') {
      classlist.push(styles.text, styles.lightText)
    } else if (type === 'secondary' || type === 'tertiary') {
      classlist.push(fonts.buttonReg, styles.darkText)
    } else {
      classlist.push(fonts.buttonReg, styles.grayText)
    }

    return classlist
  }

  const iconGen = (icon, type) => {
    let color

    if (type === 'primary' || type === 'destructive') {
      color = colors.white
    } else if (type === 'secondary' || type === 'tertiary') {
      color = colors.primary
    } else if (type === 'disabled') {
      color = colors.grey
    } else {
      color = colors.primary
    }

    return buildIcon({ name: icon, color: color, size: 24 })
  }

  const formatContent = (icon, iconPlacement, text, type) => {
    if (icon) {
      switch (iconPlacement) {
        case 'left':
          return (
            <View style={styles.contentWrapper}>
              {iconGen(icon, type)}
              <Text style={getTextStyle(type, size)}>{text}</Text>
            </View>
          )
        case 'right':
          return (
            <View style={styles.contentWrapper}>
              <Text style={getTextStyle(type, size)}>{text}</Text>
              {iconGen(icon, type)}
            </View>
          )
        case 'both':
          return (
            <View style={styles.contentWrapper}>
              {iconGen(icon, type)}
              <Text style={getTextStyle(type, size)}>{text}</Text>
              {iconGen(icon, type)}
            </View>
          )
        default:
          return <Text style={getTextStyle(type, size)}>{text}</Text>
      }
    }
  }

  const buttonMain = (
    <View
      style={[
        styles.centered,
        styles[type],
        styles[size],
        style,
        size === 'small' ? styles.buttonSmall : styles.buttonReg,
      ]}
    >
      {formatContent(icon, iconPlacement, text, type)}
    </View>
  )

  const wrappedButton = type => {
    if (type === 'disabled') {
      return buttonMain
    } else {
      return (
        <Pressable
          onPress={handlePress}
          style={({ pressed }) => [
            styles.centered,
            { opacity: pressed ? 0.8 : 1 },
          ]}
        >
          {buttonMain}
        </Pressable>
      )
    }
  }

  return wrappedButton(type)
}

Button.propTypes = {
  type: PropTypes.string.isRequired, // primary, secondary, tertiary, disabled, destructive
  size: PropTypes.string.isRequired, // fullWidth, large, medium, small
  icon: PropTypes.string.isRequired,
  iconPlacement: PropTypes.string, // left, right, both, none
  text: PropTypes.string,
  style: PropTypes.object,
  handlePress: PropTypes.func,
}

export default Button
