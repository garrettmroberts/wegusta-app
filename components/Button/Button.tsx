import { Text, View, Pressable } from 'react-native'

import styles from './styles'
import Fonts from '../../constants/Fonts'
import Colors from '../../constants/Colors'
import buildIcon from '../../utils/buildIcon'

type ButtonProps = {
  type: ButtonType,
  size: ButtonSizes,
  icon?: string,
  iconPlacement?: iconPlacementsOptions,
  text?: string,
  style?: object,
  handlePress?: () => void
}

type ButtonSizes = 'small' | 'medium' | 'large' | 'fullWidth'
type ButtonType = 'primary' |'secondary' |'tertiary' |'disabled' | 'destructive'
type iconPlacementsOptions = 'left' | 'right' | 'both' | 'none'

const Button = ({
  type,
  size,
  icon,
  iconPlacement,
  text,
  style,
  handlePress,
}: ButtonProps) => {  

  const getTextStyle = (type: ButtonType, size: ButtonSizes) => {
    const classlist = []
    classlist.push(styles.text)

    classlist.push(size === 'small' ? Fonts.buttonSmall : Fonts.buttonReg)
    if (type === 'primary' || type === 'destructive') {
      classlist.push(styles.text, styles.lightText)
    } else if (type === 'secondary' || type === 'tertiary') {
      classlist.push(Fonts.buttonReg, styles.darkText)
    } else {
      classlist.push(Fonts.buttonReg, styles.grayText)
    }

    return classlist
  }

  const iconGen = (icon: any, type: ButtonType) => {
    if (icon && iconPlacement) {
      let color
  
      if (type === 'primary' || type === 'destructive') {
        color = Colors.white
      } else if (type === 'secondary' || type === 'tertiary') {
        color = Colors.primary
      } else if (type === 'disabled') {
        color = Colors.grey
      } else {
        color = Colors.primary
      }
  
      return buildIcon({ name: icon, color: color, size: 24 })
    }
  }

  const formatContent = (type: ButtonType, text?: string, icon?: any, iconPlacement?: iconPlacementsOptions) => {
    if (icon && iconPlacement) {
      switch (iconPlacement) {
      case 'left':
        return (
          <View style={styles.contentWrapper}>
            {iconGen(icon, type)}
            <Text style={getTextStyle(type, size)}>{text}</Text>
          </View>
        )
      case 'right':
        return (
          <View style={styles.contentWrapper}>
            <Text style={getTextStyle(type, size)}>{text}</Text>
            {iconGen(icon, type)}
          </View>
        )
      case 'both':
        return (
          <View style={styles.contentWrapper}>
            {iconGen(icon, type)}
            <Text style={getTextStyle(type, size)}>{text}</Text>
            {iconGen(icon, type)}
          </View>
        )
      }
    } else {
      return <Text style={getTextStyle(type, size)}>{text}</Text>
    }

  }

  const buttonMain = (
    <View
      style={[
        styles.centered,
        styles[type],
        styles[size],
        style,
        size === 'small' ? styles.buttonSmall : styles.buttonReg,
      ]}
    >
      {formatContent(type, text, icon, iconPlacement)}
    </View>
  )

  const wrappedButton = (type: ButtonType) => {
    if (type === 'disabled') {
      return buttonMain
    } else {
      return (
        <Pressable
          onPress={handlePress}
          style={({ pressed }) => [
            styles.centered,
            { opacity: pressed ? 0.8 : 1 },
          ]}
        >
          {buttonMain}
        </Pressable>
      )
    }
  }

  return wrappedButton(type)
}

export default Button