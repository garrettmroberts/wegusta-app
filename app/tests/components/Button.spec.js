import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Text, Pressable } from 'react-native';
import Button from '../../components/Button';

describe('Button Component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Button text="Balyhoo" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Displays text prop in button', () => {
    const button = shallow(<Button text="Ballyhoo" />);
    expect(button.contains(<Text>Ballyhoo</Text>));
  });

  it('passes extra props to Pressable component', () => {
    const button = shallow(<Button text="Ballyhoo" prop_1="true" />);
    const pressable = button.find(Pressable);
    expect(pressable.props().prop_1).toEqual('true');
  });
});
