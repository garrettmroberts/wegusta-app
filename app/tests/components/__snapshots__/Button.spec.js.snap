// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "opacity": 1,
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "#6546F9",
        "borderRadius": 12,
        "height": 48,
        "justifyContent": "center",
        "width": 702,
      },
    ]
  }
  text="Balyhoo"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#6546F9",
        "borderRadius": 12,
        "height": 48,
        "justifyContent": "center",
        "width": 702,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#ffffff",
          "fontFamily": "Avenir",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "900",
          "letterSpacing": 0,
          "lineHeight": 24,
          "textAlign": "center",
        }
      }
    >
      Balyhoo
    </Text>
  </View>
</View>
`;
